DxDispatch/src/dxdispatch/Executor.cpp

```diff
@@ -188,18 +214,63 @@ void Executor::operator()(const Model::DispatchCommand& command)
     }
     PIXEndEvent();
 
-    uint32_t iterations = dispatchDurations.size();
+    m_device->GetCommandList()->ResolveQueryData(m_device->GetTimestampHeap(), D3D12_QUERY_TYPE_TIMESTAMP, 0, timestampCount, timestampReadbackBuffer.Get(), 0);
+    m_device->DispatchAndWait();
+
+    void* pData = nullptr;
+    D3D12_RANGE readRange = { 0, sizeof(uint64_t) * timestampCount };
+    timestampReadbackBuffer->Map(0, &readRange, &pData);
+
+    const uint64_t* pTimestamps = reinterpret_cast<uint64_t*>(pData);
+    timestamps.insert(timestamps.end(), pTimestamps, pTimestamps + timestampCount);
+
+
+    uint32_t iterations = dispatchDurationsCPU.size();
     // Skip the first dispatch (assuming multiple dispatches) since it warms up the pipeline.
     int skipped = (iterations > 1) ? 1 : 0;
-    double totalTime = std::accumulate(dispatchDurations.begin() + skipped, dispatchDurations.end(), 0.0);
-    double avgTime = totalTime / (dispatchDurations.size() - skipped);
+    double totalTimeCPU = std::accumulate(dispatchDurationsCPU.begin() + skipped, dispatchDurationsCPU.end(), 0.0);
+    double avgTimeCPU = totalTimeCPU / (dispatchDurationsCPU.size() - skipped);
+
+    std::sort(dispatchDurationsCPU.begin(), dispatchDurationsCPU.end());
+    double medianTimeCPU = dispatchDurationsCPU[iterations / 2];
+    double minTimeCPU = dispatchDurationsCPU[0];
+    double maxTimeCPU = dispatchDurationsCPU[iterations - 1];
+
+    uint64_t frequency;
+    m_device->GetCommandQueue()->GetTimestampFrequency(&frequency);
+
+    uint32_t samples = std::min(iterations, (uint32_t)timestamps.size() / 2);
+    std::vector<double> dispatchDurationsGPU(samples);
+
+    for (uint32_t i = 0; i < samples; ++i) {
+        uint64_t timestampDelta = (timestamps[2 * i + 1] - timestamps[2 * i]) * 1000;
+        
+        dispatchDurationsGPU[i] = double(timestampDelta) / frequency;
+    }
+
+    double totalTimeGPU = std::accumulate(dispatchDurationsGPU.begin() + skipped, dispatchDurationsGPU.end(), 0.0);
+    double avgTimeGPU = totalTimeGPU / (dispatchDurationsGPU.size() - skipped);
```

Comment:
Really minor, but I suppose there's need to `+ skipped` (line 251) and `- skipped` (line 252) if the number of iterations > timestamps.size(). In this case you'll already have overwritten the first dispatch timestamps. This only applies to the GPU timings, of course.